sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -787,6 +788,28 @@ object BooleanSimplification extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 /**
+ * Simplifies semantically-equal non-nullable expressions:
+ * 1) Replace '=', '<=', and '>=' with 'true' literal
+ * 2) Replace '<' and '>' with 'false' literal
+ */
+object NonNullableBinaryComparisonSimplification extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsUp {
+      // True with equality
+      case a EqualTo b if !a.nullable && !b.nullable && a.semanticEquals(b) => TrueLiteral
+      case a EqualNullSafe b if !a.nullable && !b.nullable && a.semanticEquals(b) => TrueLiteral
```

Comment:
Oh. @cloud-fan . When I looked at the `EqualNullSafe` one more time, I realized that that `a EqualNullSafe b` is null-safe but both `a` and `b` are not. The following is the code snippet of `EqualNullSafe`. 
```
case class EqualNullSafe(left: Expression, right: Expression) extends BinaryComparison {
...
  override def nullable: Boolean = false
...
  override def eval(input: InternalRow): Any = {
    val input1 = left.eval(input)
    val input2 = right.eval(input)
    if (input1 == null && input2 == null) {
      true
```
In this case, I think we should keep `!a.nullable && !b.nullable` because we need to evaluate `a` and `b` indeed. How do you think about that?